### PR TITLE
gh-100690: Prevent prefix "called_" for attributes of safe mock objects

### PR DESCRIPTION
**Pull request by cklein**
Prevent prefix "called_" for mocks in safe mode

- Issue: gh-100690

### CODE REVIEW

**Review by cjw296 — COMMENTED** [review]
I'm -0 on this change, feels like we're into the diminishing realms where someone out there is going to have a legitimate use case for `called_name` or some such and this then prevents them using a safe mock.

Who else is hitting this?

**Comment by cklein** [comment]
@cjw296, the same applies to any of the forbidden prefixes. This could be solved by allowing those prefixes with a spec. Currently, even if the name appears in `_mock_methods` (https://github.com/python/cpython/blob/main/Lib/unittest/mock.py#L651 evaluates to true), an error will be raised. But I guess this could be a different and more fundamental discussion.

**Comment by cklein** [comment]
I have been thinking a bit more about the whole problem. It seems the current behavior does not match the error message:

```
>>> class Foo:
...     def assert_bar(self):
...         pass
...
>>> from unittest.mock import Mock
>>> m = Mock(spec=Foo)
>>> m.assert_bar()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/mock.py", line 648, in __getattr__
    raise AttributeError(
AttributeError: 'assert_bar' is not a valid assertion. Use a spec for the mock if 'assert_bar' is meant to be an attribute.
```

The error message says "_Use a spec for the mock if 'assert_bar' is meant to be an attribute._",  but this example should pass according to the error message. There's a spec and the name `assert_bar` should be supported.

With https://github.com/cklein/cpython/commit/37aa2918f706e1c3333ac5be4c0628c9f1201d2e applied, it behaves like described:

```
>>> class Foo:
...     def assert_bar(self):
...         pass
...
>>> from unittest.mock import Mock
>>> m = Mock(spec=Foo)
>>> m.assert_bar()
<Mock name='mock.assert_bar()' id='4384202080'>
>>> m.assert_frobnicated()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/cklein/github/cklein/cpython/Lib/unittest/mock.py", line 652, in __getattr__
    raise AttributeError("Mock object has no attribute %r" % name)
AttributeError: Mock object has no attribute 'assert_frobnicated'
```

**Comment by merwok** [comment]
The prefixes checked could be more specific like `called_once_with`, `called_once`, `called_with` to alleviate the concern about forbidding too much.

**Comment by cjw296** [comment]
@cklein - I think you're right about this being a complicated issue, perhaps a PR isn't the best place to discuss it? Could you maybe close this PR an open an issue explaining your concerns and some possible ways forward?

**Comment by AlexWaygood** [comment]
> Could you maybe close this PR an open an issue explaining your concerns and some possible ways forward?

OP has already done so:

- They first opened a discussion on discuss.python.org, https://discuss.python.org/t/include-prefix-called-in-list-of-forbidden-method-prefixes-for-mock-objects-in-unsafe-mode/22249.
- After discussing it there, they were advised by a core developer to open an issue: https://github.com/python/cpython/issues/100690
- And a PR (this one).

**Comment by cklein** [comment]
I can also create another issue for my second branch about the spec not being properly checked, if that helps

**Comment by cklein** [comment]
I'm aware that the documentation is now off. I'd like to get some feedback on the change first, if that's OK

File: Lib/unittest/mock.py
Comment (cjw296) [inline Lib/unittest/mock.py]: Personally I dislike these kind of bug links, the comment is fine without it, and in a decade's time when we're no longer using GitHub for issue tracking, this will be a wart ;-)

File: Lib/unittest/mock.py
Comment (cjw296) [inline Lib/unittest/mock.py:1066]: Much prefer this approach! 🎉

**Comment by cjw296** [comment]
Approach is good, so once the docs are fixed up, should be good to land.
Would a backport release once that's done be helpful? (I mean, I'll probably crank the handle anyway, but thought I'd ask!)

File: Lib/unittest/mock.py
Comment (merwok) [reply Lib/unittest/mock.py]: Same for me

File: Misc/NEWS.d/next/Library/2023-01-02-16-59-49.gh-issue-100690.2EgWPS.rst
Comment (merwok) [inline Misc/NEWS.d/next/Library/2023-01-02-16-59-49.gh-issue-100690.2EgWPS.rst]: Minor thing: can you wrap to 72 columns?  And add markup to Mock and AttributeError, plus code markup for the prefixes on last line.

**Comment by cklein** [comment]
> Would a backport release once that's done be helpful?

Personally I'd like to see it in earlier versions of Python.
But then I'm a bit torn because on the one hand this will definitely make some tests fail, on the other hand those tests will finally be properly executed and some unwanted behavior might be uncovered.

**Comment by cjw296** [comment]
@cklein - 3.10 and 3.11 are open for bug fixes so their next third point release would contain this once its landed.

**Review by cjw296 — APPROVED** [review]

**Comment by cjw296** [comment]
The Azure Pipelines failure seems unrelated.

File: Lib/unittest/mock.py
Comment (sobolevn) [inline Lib/unittest/mock.py:1066]: Sorry for being late to the party, but anyways :)

I think that `ATTRIB_DENY_LIST` is an implementation detail, so it should be `_ATTRIB_DENY_LIST`. And also, it is not suited for mutation, isn't it? So, it should be a `frozenset` instead :)

Comment (sobolevn) [reply Lib/unittest/mock.py:1066]: Since this PR is already merged, I've opened https://github.com/python/cpython/pull/100819 with this proposal. Feel free to close if it should be public and mutable!

Comment (merwok) [reply Lib/unittest/mock.py:1066]: It is already non-public by not being included in `__all__`, and the mutability is not really an issue.

**Pull request merged**
